internal/middleware/perfinsights/perfinsights.go

```diff
@@ -0,0 +1,189 @@
+package perfinsights
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/ccoveille/go-safecast"
+	"github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/rs/zerolog/log"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+
+	"github.com/authzed/ctxkey"
+	"github.com/authzed/grpcutil"
+
+	"github.com/authzed/spicedb/pkg/spiceerrors"
+)
+
+type APIShapeLabel string
+
+const (
+	ResourceTypeLabel     APIShapeLabel = "resource_type"
+	ResourceRelationLabel APIShapeLabel = "resource_relation"
+	SubjectTypeLabel      APIShapeLabel = "subject_type"
+	SubjectRelationLabel  APIShapeLabel = "subject_relation"
+	NameLabel             APIShapeLabel = "name"
+	FilterLabel           APIShapeLabel = "filter"
```

Comment:
Please note that, [while Native Histograms need only 1 time series to represent a dynamic set of buckets](https://prometheus.io/docs/concepts/metric_types/#histogram) (massive improvement over traditional histograms!) they still have the same cadinality explosion problem for other labels: for very large schemas, this can lead to a _many_ combination of labels and thus number of timeseries.

This is worth the price, and making perf insights optional is the right move so that users aren't surprised when the _[insert observability vendor here]_ bill comes.